### ISSUE TITLE
"docker logs -f xxx" lost last log if no newline at the end

### ISSUE THREAD

**Issue by HunterXuan** [moby/moby#43860]
### Description

When I run a container on linux system which prints something without newline (\n) at the end, the last log would be lost if using `docker logs -f` command to follow its logs.

And such behavior differs with different operate system:

1. On CentOS, the last log is always lost
2. On Windows, the last log is lost sometimes
3. On Mac (M1 chip), the last log is printed as expected

### Reproduce

1. run a container that echo something without a newline at the end

```bash
docker run --name test --rm  alpine /bin/sh -c "sleep 5; echo hello; echo -n world"
```

2. use `docker logs -f` to follow the logs, and you will never get the "world" output

```bash
docker logs -f test
```

### Expected behavior

`docker logs -f` should print the last log even if no newline at the end.

### docker version

```bash
Client: Docker Engine - Community
 Version:           20.10.8
 API version:       1.41
 Go version:        go1.16.6
 Git commit:        3967b7d
 Built:             Fri Jul 30 19:55:49 2021
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.8
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.16.6
  Git commit:       75249d8
  Built:            Fri Jul 30 19:54:13 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.9
  GitCommit:        e25210fe30a0a703442421b0f60afac609f950a3
 runc:
  Version:          1.0.1
  GitCommit:        v1.0.1-0-g4144b63
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Build with BuildKit (Docker Inc., v0.6.1-docker)
  scan: Docker Scan (Docker Inc., v0.8.0)

Server:
 Containers: 4
  Running: 4
  Paused: 0
  Stopped: 0
 Images: 5
 Server Version: 20.10.8
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: active
  NodeID: v993fmlrkssle7n7x5hfjcg5p
  Is Manager: true
  ClusterID: ih8yo8lb4npcfa56ij8e4en1j
  Managers: 1
  Nodes: 3
  Default Address Pool: 10.0.0.0/8  
  SubnetSize: 24
  Data Path Port: 4789
  Orchestration:
   Task History Retention Limit: 5
  Raft:
   Snapshot Interval: 10000
   Number of Old Snapshots to Retain: 0
   Heartbeat Tick: 1
   Election Tick: 10
  Dispatcher:
   Heartbeat Period: 5 seconds
  CA Configuration:
   Expiry Duration: 3 months
   Force Rotate: 0
  Autolock Managers: false
  Root Rotation In Progress: false
  Node Address: 10.0.0.17
  Manager Addresses:
   10.0.0.17:2377
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: e25210fe30a0a703442421b0f60afac609f950a3
 runc version: v1.0.1-0-g4144b63
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 3.10.0-1160.36.2.el7.x86_64
 Operating System: CentOS Linux 7 (Core)
 OSType: linux
 Architecture: x86_64
 CPUs: 1
 Total Memory: 1.795GiB
 Name: web-qq-01
 ID: ZLO2:2LRS:HRY7:3OT7:2K3L:GW63:QDCW:WHWJ:OOSK:5OW2:3R3F:Y5BK
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by thaJeztah** [comment]
Thanks for reporting.


Hm, so, _in general_, I think this is somewhat expected behaviour; docker's logging drivers are line delimited, which means that a log entry is either created;

- if a newline is encountered in the output
- or, if the buffer size is reached (in which case a "partial" (incomplete) log message is created; https://github.com/moby/moby/blob/a8a769f04f7ae44117772943ef0f4607b54e05ff/daemon/logger/copier.go#L19-L21


I think the scenario you're describing is where a container is exiting while it was still printing output; that last output did not contain a newline, and therefore (from the logging driver's perspective) not yet "complete", and therefore not a log entry.

Perhaps we _could_ try to flush the buffer when the container exits, but even if we did, I _think_ there still would be a race condition between trying to flush the buffer and stopping and (in case of `--rm`) removing the container. (Logs are no longer collected when the container exits).

> On Windows, the last log is lost sometimes

.. and based on this comment (I haven't followed all code paths), it's possible that flushing already happens, but due to the race "may" or "may not" succeed.

I'm not sure about these differences though;

> On Windows, the last log is lost sometimes
> On Mac (M1 chip), the last log is printed as expected

My best guess here would be that stopping and removing the container may be slower on Windows and possibly flushing the buffer is more likely to "work".

./cc @corhere perhaps you have some ideas (I know you looked into some of the logging code recently)

**Comment by corhere** [comment]
> Perhaps we _could_ try to flush the buffer when the container exits, but even if we did, I _think_ there still would be a race condition between trying to flush the buffer and stopping and (in case of `--rm`) removing the container.

As far as the interprocess communication is concerned, I don't think a race between a process writing to an STDIO stream and exiting is supposed to be possible on Linux, notwithstanding kernel bugs or deliberate POSIX noncompliance:

> POSIX requires that a [`read(2)`](https://man7.org/linux/man-pages/man2/read.2.html) that can be proved to occur after a `write()` has returned will return the new data.
> -- [write(2), Linux Programmer's Manual](https://man7.org/linux/man-pages/man2/write.2.html)

Observing a process exit certainly qualifies as proof that any `write()` calls it had made had returned. Furthermore, the container's STDIO streams are connected to docker's logging subsystem over FIFOs, for which there is a mechanism for the read side to know when the writer is gone and buffer drained:

> If all file descriptors referring to the write end of a pipe have been closed, then an attempt to [`read(2)`](https://man7.org/linux/man-pages/man2/read.2.html) from the pipe will see end-of-file ([`read(2)`](https://man7.org/linux/man-pages/man2/read.2.html) will return 0).
> -- [pipe(7), Linux Programmer's Manual](https://man7.org/linux/man-pages/man7/pipe.7.html)

So as far as I can tell, there is no technical limitation with the Linux syscall interface which would prevent us from reliably reading every last byte the container processes write to STDOUT or STDERR. Any race conditions are therefore most likely entirely in userspace.

> it's possible that flushing already happens, but due to the race "may" or "may not" succeed.

I can confirm: the code which breaks the STDOUT and STDERR streams into lines does flush any remaining partial line. https://github.com/moby/moby/blob/98d8343aa28de0d499464b5529e6b8ccc92e9313/daemon/logger/copier.go#L134-L136

> 
> I'm not sure about these differences though;
> 
> > On Windows, the last log is lost sometimes
> > On Mac (M1 chip), the last log is printed as expected
> 
> My best guess here would be that stopping and removing the container may be slower on Windows and possibly flushing the buffer is more likely to "work".

I can confirm that stopping the container is slower on Windows. The builtin Windows runtime waits until the process has exited _and_ the runtime has finished shutting the "compute system" down before informing the daemon that the process has exited.

> ./cc @corhere perhaps you have some ideas (I know you looked into some of the logging code recently)

I was able to reproduce the issue on Docker Desktop for Mac (amd64) version 4.9.1 (81317) on my first try, so I suspect that platform differences only change the likelihood of the race settling one way or the other. I am also able to confirm that container auto-delete is not necessary to reproduce the issue. Following logs while the container is running results in the last partial log line being emitted by the `docker run` command but sometimes not a concurrent `docker logs -f` command. Running `docker logs test` after the container exits reveals that the partial log message _has been successfully flushed_ in all cases, which indicates that the root cause has to do with following a container's logs, not capturing or writing them.

I think I already found the race condition and fixed it in #43294, first in https://github.com/moby/moby/pull/43294/commits/3844d1a3d16be5a4e2e6ddf8c95afad6ad6bfd03 and then in the total rewrite of the follower. I was unable to replicate the issue in ten attempts with the latest HEAD so I believe the bug has been fixed for v22.06. That first fix is small and targeted enough to potentially be backported to v20.10.

**Comment by HunterXuan** [comment]
> I was unable to replicate the issue in ten attempts with the latest HEAD so I believe the bug has been fixed for v22.06. That first fix is small and targeted enough to potentially be backported to v20.10.

Looking forward to the fix in v20.10

**Comment by kaffarell** [comment]
Yep, tried this a few times, works flawlessly on latest HEAD. I think we can close this issue @thaJeztah

**Comment by thaJeztah** [comment]
Thanks! Let me close this one as fixed through https://github.com/moby/moby/pull/43294. That patch will be part of the next (22.x) release, but won't be in a 20.10 patch release (patch is too large to be considered for a patch release).
